backends/ubpf/ubpfTable.cpp

```diff
@@ -0,0 +1,412 @@
+            const IR::P4Action *action;
+            cstring valueName;
+
+        public:
+            UbpfActionTranslationVisitor(cstring valueName, const UBPFProgram *program) :
+                    EBPF::CodeGenInspector(program->refMap, program->typeMap), program(program),
+                    action(nullptr), valueName(valueName) {
+                CHECK_NULL(program);
+            }
+
+            bool preorder(const IR::PathExpression *expression) override {
+                auto decl = program->refMap->getDeclaration(expression->path, true);
+                if (decl->is<IR::Parameter>()) {
+                    auto param = decl->to<IR::Parameter>();
+                    bool isParam = action->parameters->getParameter(param->name) == param;
+                    if (isParam) {
+                        builder->append(valueName);
+                        builder->append("->u.");
+                        cstring name = EBPF::EBPFObject::externalName(action);
+                        builder->append(name);
+                        builder->append(".");
+                        builder->append(expression->path->toString());  // original name
+                        return false;
+                    }
+                }
+
+                visit(expression->path);
+                return false;
+            }
+
+            bool preorder(const IR::MethodCallExpression *expression) override {
+                auto mi = P4::MethodInstance::resolve(expression, refMap,
+                                                      typeMap);
+                auto ef = mi->to<P4::ExternFunction>();
+                if (ef != nullptr) {
+                    if (ef->method->name.name == program->model.drop.name) {
+                        builder->appendFormat("%s = false", program->control->passVariable);
+                        return false;
+                    } else if (ef->method->name.name == program->model.pass.name) {
+                        builder->appendFormat("%s = true", program->control->passVariable);
+                        return false;
+                    } else if (ef->method->name.name ==
+                               program->model.ubpf_time_get_ns.name) {
+                        builder->emitIndent();
+                        builder->append(program->model.ubpf_time_get_ns.name + "()");
+                        return false;
+                    }
+                }
+                program->control->codeGen->preorder(expression);
+                return false;
+            }
+
+            void convertActionBody(const IR::Vector<IR::StatOrDecl> *body) {
+                for (auto s : *body) {
+                    if (!s->is<IR::Statement>()) {
+                        continue;
+                    } else if (auto block = s->to<IR::BlockStatement>()) {
+                        convertActionBody(&block->components);
+                        continue;
+                    } else if (s->is<IR::ReturnStatement>()) {
```

Comment:
Do you want to put here a BUG_CHECK that no return or exit is encountered?

Reply:
I shouldn't put BUG_CHECK here, because retrun or exit could be encountered here (as both exit and return are allowed for actions according to specs). However, we use this `else if` block to implement a proper behavior, i.e. don't emit any statements following exit or return.